### PR TITLE
shapeworks Python module 

### PR DESCRIPTION
**Pull request by archanasri**

### CODE REVIEW

**Comment by sheryjoe** [comment]
@archanasri could you add instructions on how to test this PR?

**Comment by cchriste** [comment]
Look at all those green checks!!! This is fantastic. Plus all the tests actually run and pass, and there is a robust collection of tests for the Image class, the Python module, and the shapeworks executable, and at least one or more tests for Mesh, Particles, and Optimize.

**Comment by akenmorris** [comment]
Could we remove the test output from this branch?  (Still need to change the output location of the tests #637)

**Comment by cchriste** [comment]
> Could we remove the test output from this branch? (Still need to change the output location of the tests #637)
Good idea. Whoever works on that issue would be welcome to address it here. It's currently completely up to date with master.

> As an aside, I would recommend against doing blanket additions of any new files.
Can you please describe "blanket additions"?

**Comment by akenmorris** [comment]
> Whoever works on that issue

I don't think anyone is working on it.  Are you volunteering? ;)

> Can you please describe "blanket additions"?

I mean "git add --all" or whatever it is people are doing that ends up adding all of the test output into git.

**Review by akenmorris — CHANGES_REQUESTED** [review]
Let's remove the test output (for example Testing/data/use_normals/output/*)

**Comment by akenmorris** [comment]
Hey, is there any particular reason the Python API should be optional?

**Comment by archanasri** [comment]
> Hey, is there any particular reason the Python API should be optional?

Nope. There is no reason for it to be optional.

**Comment by sheryjoe** [comment]
@HeavenlyBerserker @iyerkrithika21 
- Going over *Utils.py and make sure shapeworks command calls are replaced with Python APIs and document any issues
- Run use cases and compare the output with the master branch (means shape, modes of variation, and individual sample
- Put together a simple Jupyter notebook (or a python script) that mimics a simple use case (e.g. ellipsoid use case) 

Please report all types of issues that you face, especially usability.

**Comment by akenmorris** [comment]
Can we reorganize the python tests so that they can be run individually?

when I run:

PythonTests --gtest_filter=pythonTests.resize

It doesn't work (can't find resize.py).

This makes it hard to work on one test (at the end) if you have to wait and watch all the other tests complete.

**Review by akenmorris — APPROVED** [review]

**Comment by cchriste** [comment]
> Can we reorganize the python tests so that they can be run individually?

Good call. I created #822 since this relates to more than just the Python module and can be addressed independently.

**Comment by akenmorris** [comment]
What else needs to be done to get this merged?  I've tried it and given my approval.

**Comment by sheryjoe** [comment]
> @HeavenlyBerserker @iyerkrithika21
> 
> * Going over *Utils.py and make sure shapeworks command calls are replaced with Python APIs and document any issues
> * Run use cases and compare the output with the master branch (means shape, modes of variation, and individual sample
> * Put together a simple Jupyter notebook (or a python script) that mimics a simple use case (e.g. ellipsoid use case)
> 
> Please report all types of issues that you face, especially usability.

@HeavenlyBerserker @iyerkrithika21  what is the status of your test/review?

**Comment by sheryjoe** [comment]
@cchriste @archanasri @akenmorris @iyerkrithika21 
I am getting `ModuleNotFoundError: No module named 'shapeworks'` when running the ellipsoid use case. I did a fresh conda_installs, build_dependencies, cmake, and make. The build_dependencies worked but with the below error at the end.

./build_dependencies.sh: line 282: FINDQT: command not found
./build_dependencies.sh: line 284: FINDQT: command not found

**Comment by akenmorris** [comment]
> FINDQT: command not found

This should be fixed now if you pull the latest.

**Comment by iyerkrithika21** [comment]
WIP on the notebook : [https://github.com/SCIInstitute/ShapeWorks/blob/pybind_test/Examples/Python/Ellipsoid_Use_Case.ipynb](url)
As discussed yesterday with @cchriste and @HeavenlyBerserker , another notebook will be created which does not use any pre-defined functions from GroomUtils.

**Comment by sheryjoe** [comment]
Thanks @akenmorris 

I still have `ModuleNotFoundError: No module named 'shapeworks'` @cchriste @archanasri 
I started from a clean anaconda install ...

**Comment by cchriste** [comment]
> Thanks @akenmorris
> 
> I still have `ModuleNotFoundError: No module named 'shapeworks'` @cchriste @archanasri
> I started from a clean anaconda install ...

Please see the conversation we had last night on #dev from slack.

**Comment by sheryjoe** [comment]
> > Thanks @akenmorris
> > I still have `ModuleNotFoundError: No module named 'shapeworks'` @cchriste @archanasri
> > I started from a clean anaconda install ...
> 
> Please see the conversation we had last night on #dev from slack.

Adding the fix here for future as one might get lost digging into the slack channel. 

Need to set PYTHONPATH (see below) -- **shouldn't that be already added to RunUseCase?** at this point we can't wait for the long standing startsw script for this fix. I wouldn't expect users to set this path. 

PYTHONPATH, or open python (or a notebook) and insert/append to sys.path
It’s necessary no matter what, and one of the reasons we talked about having a startsw script (the other reason was to put the shapeworks executable in their path).

If you open python or a notebook:
```
import sys
sys.path.insert(0, /path/to/build/bin)
```

if you do it before that:
```
export PYTHONPATH=/path/to/build/bin:$PYTHONPATH
```

**Comment by cchriste** [comment]
Thanks for copying the instructions I shared on slack last night.

Yes, RunUseCase already sets `sys.path`, but it is blind batch processing,
which misses the point (and power) of the shapeworks Python module.

The proposed `startsw` script (called `setupenv` in the [upgrade_deps](
https://github.com/SCIInstitute/ShapeWorks/tree/upgrade_deps) branch) gets
the user part way to having the `shapeworks` executable (and the others) in
their PATH, the Python module in their PYTHONPATH, and the libs in the
LD_LIBRARY_PATH, customized for whichever platform needs whatever envvar.
There were issues creating it that are described in detail in [issue 478](
https://github.com/SCIInstitute/ShapeWorks/issues/478).

**Comment by sheryjoe** [comment]
@jadie1 left atrium use case segfaults when running it with subsample

```
(shapeworks) shireen$ python RunUseCase.py --use_case left_atrium  --use_subsample --num_subsample 5

Step 1. Extract Data

Segmentation fault: 11
```

**Comment by sheryjoe** [comment]
@jadie1 @iyerkrithika21 

left atrium use case results in the below error in cropping when running tiny test. 

############## Cropping ##############
Traceback (most recent call last):
  File "RunUseCase.py", line 75, in <module>
    module.Run_Pipeline(args)
  File "XX/Examples/Python/left_atrium.py", line 173, in Run_Pipeline
    croppedFiles = applyCropping(groomDir + "cropped", rigidFiles, groomDir + "aligned/*.aligned.nrrd")
  File "XX/Examples/Python/GroomUtils.py", line 258, in applyCropping
    region = ImageUtils.boundingBox(glob.glob(initPath+"/*.nrrd"))
ValueError: Image sizes do not match (Output/left_atrium/groomed/aligned/CARMA0154.laendo_no_veins.isores.center.pad.com.aligned.nrrd)

**Comment by sheryjoe** [comment]
@jadie1 @iyerkrithika21 

same thing for the running the full use case

############## Cropping ##############
Traceback (most recent call last):
  File "RunUseCase.py", line 75, in <module>
    module.Run_Pipeline(args)
  File "XX/Examples/Python/left_atrium.py", line 173, in Run_Pipeline
    croppedFiles = applyCropping(groomDir + "cropped", rigidFiles, groomDir + "aligned/*.aligned.nrrd")
  File "XX/Examples/Python/GroomUtils.py", line 258, in applyCropping
    region = ImageUtils.boundingBox(glob.glob(initPath+"/*.nrrd"))
ValueError: Image sizes do not match (Output/left_atrium/groomed/aligned/CARMA1309.laendo_no_veins.isores.center.pad.com.aligned.tpSmoothDT.nrrd)

**Comment by cchriste** [comment]
Isn't the tiny test of all the Examples/Python use cases in the github
action tests?
If not, is there any reason not to add them for everything?

**Comment by akenmorris** [comment]
> Isn't the tiny test of all the Examples/Python use cases in the github action tests? If not, is there any reason not to add them for everything?

Yes, the problem is that femur --tiny_test requires downloading a 10GB zip file from the data portal.  Left Atrium case is large as well.  I'm not sure which ones are self contained besides ellipsoid.

**Comment by jadie1** [comment]
@sheryjoe I am getting this error when I run the ellipsoid tiny test too, are you seeing that? @akenmorris It is passing the ellipsoid test right? So this error isn't occurring for the github build?

**Comment by iyerkrithika21** [comment]
> 
> 
> @sheryjoe I am getting this error when I run the ellipsoid tiny test too, are you seeing that? @akenmorris It is passing the ellipsoid test right? So this error isn't occurring for the github build?

The first time I tried tiny_test on ellipsoid, it failed. Then I tried with sub_samples(5 and 3), it passed after which I tried tiny_test and it is working now

**Comment by akenmorris** [comment]
> It is passing the ellipsoid test right? So this error isn't occurring for the github build?

Yes, it works there and it works for me as long as I set PYTHONPATH to my bin directory.  Yours still can't find the shapeworks module even with PYTHONPATH set?  Is that right?

**Comment by jadie1** [comment]
> > It is passing the ellipsoid test right? So this error isn't occurring for the github build?
> 
> Yes, it works there and it works for me as long as I set PYTHONPATH to my bin directory. Yours still can't find the shapeworks module even with PYTHONPATH set? Is that right?

Importing shapeworks works if I set the PYTHONPATH. I just also have to set the LD_LIBRARY_PATH or it can't find libitkgdcmjpeg8-5.0.so.1.  The error I was seeing with ellipsoids was on the cropping step but I think I was using v1 ellipsoids (the new ones which are less groomed) it actually runs for me with v0 ellipsoids. I think the bug is probably in GroomUtils

**Comment by archanasri** [comment]
> Isn't the tiny test of all the Examples/Python use cases in the github action tests? If not, is there any reason not to add them for everything?

I don't think the github action tests include Examples/Python.

**Review by sheryjoe — CHANGES_REQUESTED** [review]
Please address the below issues: 

#833 
#831
#830
#828 
#827 
#825

**Comment by sheryjoe** [comment]
> > Isn't the tiny test of all the Examples/Python use cases in the github action tests? If not, is there any reason not to add them for everything?
> 
> Yes, the problem is that femur --tiny_test requires downloading a 10GB zip file from the data portal. Left Atrium case is large as well. I'm not sure which ones are self contained besides ellipsoid.

Can we download a subset? We discussed adding this functionality to the Datasets package. @0leks @akenmorris

**Comment by 0leks** [comment]
@sheryjoe can download subset by provide a list of filenames to the DatasetUtils.downloadDataset function: [http://sciinstitute.github.io/ShapeWorks/dev/datasets.html](http://sciinstitute.github.io/ShapeWorks/dev/datasets.html)
image

**Comment by sheryjoe** [comment]
> @sheryjoe can download subset by provide a list of filenames to the DatasetUtils.downloadDataset function: http://sciinstitute.github.io/ShapeWorks/dev/datasets.html
> image

This is what I recall. thanks @0leks 

@jadie1 @iyerkrithika21 we should update the use cases to only download the needed samples in case of tiny_test and use_subsample

**Comment by sheryjoe** [comment]
@akenmorris @medakk 

The below notebook sets up shapeworks python environment and tests the setup. I tested it on Mac (Big Sur) and works. Could you please test it on Linux and Windows and let me know what changes to be made to make it cross platform?

https://github.com/SCIInstitute/ShapeWorks/blob/pybind/Examples/Python/notebooks/tutorials/setting-up-shapeworks-environment.ipynb

**Comment by sheryjoe** [comment]
@akenmorris Is github actions disabled for this PR? I need the links for the artifacts after the latest commit to the pybind branch.

**Comment by akenmorris** [comment]
> Is github actions disabled for this PR? I need the links for the artifacts after the latest commit to the pybind branch.

They are enabled.  If you look at the actions page, there are several queued for this branch.  There seems to be something wrong with GitHub Actions and/or Azure Pipelines today.  Things are getting stuck, when I cancel, they come back with success.  The timings are all off.

**Comment by sheryjoe** [comment]
@akenmorris Is it possible to have links for users to download the latest artifacts (windows, linux and mac) for this PR?

**Comment by sheryjoe** [comment]
@cchriste @akenmorris @archanasri @jadie1 @medakk @iyerkrithika21 

I have finished the first wave of getting started notebooks. I have tested these on Mac Big Sur. Could you please (1) run/test the below notebooks on Catalina, Window and Linux, and (2) go over each notebook and check out the logical flow and explanation? 

- Examples/Python/notebooks/tutorials/getting-started-with-jupyter-notebooks.ipynb
- Examples/Python/notebooks/tutorials/setting-up-shapeworks-environment.ipynb
- Examples/Python/notebooks/tutorials/getting-started-with-segmentations.ipynb  
- Examples/Python/notebooks/tutorials/getting-started-with-meshes.ipynb
- Examples/Python/notebooks/tutorials/getting-started-with-exploring-segmentations.ipynb
- Examples/Python/notebooks/tutorials/getting-started-with-grooming-segmentations.ipynb

**Comment by jadie1** [comment]
@sheryjoe I have checked all of the notebooks you listed on Linux. Here are my notes:

- We could make it so the data is automatically downloaded once they have an account as it is for the use cases
- getting-started-with-grooming-segmentations.ipynb is breaking for me on the resampling step with the latest ellipsoid-v2 data
image
- The documentation looks good, I did get a couple link not found warning when I ran mkdocs serve though
- Overall I think these are very thorough and will be very helpful for uses

**Comment by sheryjoe** [comment]
> * getting-started-with-grooming-segmentations.ipynb is breaking for me on the resampling step with the latest ellipsoid-v2 data
Did you pull and build the latest changes in pybind? from the image, I think this is an old build of the pybind branch 

> * The documentation looks good, I did get a couple link not found warning when I ran mkdocs serve though
Which links?

File: Libs/Mesh/Mesh.h [hunk 1]
```diff
@@ -36,6 +36,21 @@ class Mesh
...
+  vtkIdType numVertices() const { return mesh->GetNumberOfVerts(); }
+
+  /// number of faces
+  vtkIdType numFaces() const { return mesh->GetNumberOfCells(); }
+
+  /// get bounds
+  double* bounds() const {
+    double bb[6];
+    mesh->GetBounds(bb);
+    return bb;
```
Comment (akenmorris) [inline Libs/Mesh/Mesh.h:51]: I see this warning in the build:

warning: address of stack memory associated with local variable 'bb' returned [-Wreturn-stack-address]
    return bb;

This array, 'bb' is allocated on the stack and will be gone as soon as the function returns.

Comment (archanasri) [reply Libs/Mesh/Mesh.h:51]: I have fixed this in the mesh_party branch. I created a Mesh::Region so that this does not happen.

**Pull request merged**